build/org.eclipse.cdt.build.gcc.core/src/org/eclipse/cdt/build/gcc/core/GCCToolChain.java

```diff
@@ -640,42 +645,102 @@ public String[] getCompileCommands(ILanguage language) {
 		}
 	}
 
+	/**
+	 * Construct regular expression to find any file extension for C or C++. Returns
+	 * expression shaped in form of "((cpp)|(c++)|(c))".
+	 * Copied from {@link AbstractLanguageSettingsOutputScanner}.
+	 *
+	 * @return regular expression for searching C/C++ file extensions.
+	 */
+	private String getPatternFileExtensions() {
+		IContentTypeManager manager = Platform.getContentTypeManager();
+
+		Set<String> fileExts = new HashSet<>();
+
+		IContentType contentTypeCpp = manager.getContentType(CCorePlugin.CONTENT_TYPE_CXXSOURCE);
+		fileExts.addAll(Arrays.asList(contentTypeCpp.getFileSpecs(IContentType.FILE_EXTENSION_SPEC)));
+
+		IContentType contentTypeC = manager.getContentType(CCorePlugin.CONTENT_TYPE_CSOURCE);
+		fileExts.addAll(Arrays.asList(contentTypeC.getFileSpecs(IContentType.FILE_EXTENSION_SPEC)));
+
+		String pattern = expressionLogicalOr(fileExts);
+
+		return pattern;
+	}
+
+	/**
+	 * Helper method to construct logical "or" to be used inside regular
+	 * expressions.
+	 * Copied from {@link AbstractLanguageSettingsOutputScanner}.
+	 */
+	@SuppressWarnings("nls")
+	private static String expressionLogicalOr(Set<String> fileExts) {
+		String pattern = "(";
+		for (String ext : fileExts) {
+			if (pattern.length() != 1)
+				pattern += "|";
+			pattern += "(" + Pattern.quote(ext) + ")";
+			ext = ext.toUpperCase();
+			if (!fileExts.contains(ext)) {
+				pattern += "|(" + Pattern.quote(ext) + ")";
+			}
+		}
+		pattern += ")";
+		return pattern;
+	}
+
+	private static String getFileExtension(String fileName) {
+		String ext = ""; //$NON-NLS-1$
+
+		int i = fileName.lastIndexOf('.');
+		if (i > 0) {
+			ext = fileName.substring(i + 1);
+		}
+		return ext;
+	}
+
 	@Override
 	public IResource[] getResourcesFromCommand(List<String> cmd, URI buildDirectoryURI) {
 		// Start at the back looking for arguments
 		List<IResource> resources = new ArrayList<>();
 		IWorkspaceRoot root = ResourcesPlugin.getWorkspace().getRoot();
+		String extensionPattern = getPatternFileExtensions();
```

Comment:
Below I recommend getting rid of the regular expression entirely, but if it is needed: Because this code can be processing large amounts of output we should be pre-compiling, or compiling as early as possible, the patterns.